stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+{
+	return 0.0;
+}
+
+double StatsBinner::upperBoundToFloat(const StatsBin &bin) const
+{
+	return 0.0;
+}
+
+bool StatsBinner::preferBin(const StatsBin &bin) const
+{
+	return true;
+}
+
+// Default implementation for discrete variables: there are no bins between discrete bins.
+std::vector<StatsBinPtr> StatsBinner::bins_between(const StatsBin &bin1, const StatsBin &bin2) const
+{
+	return {};
+}
+
+QString StatsVariable::unitSymbol() const
+{
+	return {};
+}
+
+QString StatsVariable::diveCategories(const dive *) const
+{
+	return QString();
+}
+
+int StatsVariable::decimals() const
+{
+	return 0;
+}
+
+double StatsVariable::toFloat(const dive *d) const
+{
+	return invalid_value<double>();
+}
+
+QString StatsVariable::nameWithUnit() const
+{
+	QString s = name();
+	QString symb = unitSymbol();
+	return symb.isEmpty() ? s : QStringLiteral("%1 [%2]").arg(s, symb);
+}
+
+QString StatsVariable::nameWithBinnerUnit(const StatsBinner &binner) const
+{
+	QString s = name();
+	QString symb = binner.unitSymbol();
+	return symb.isEmpty() ? s : QStringLiteral("%1 [%2]").arg(s, symb);
+}
+
+const StatsBinner *StatsVariable::getBinner(int idx) const
+{
+	std::vector<const StatsBinner *> b = binners();
+	if (b.empty())
+		return nullptr;
+	return idx >= 0 && idx < (int)b.size() ? b[idx] : b[0];
+}
+
+std::vector<StatsOperation> StatsVariable::supportedOperations() const
+{
+	return {};
+}
+
+// Attn: The order must correspond to the StatsOperation enum
+static const char *operation_names[] = {
+	QT_TRANSLATE_NOOP("StatsTranslations", "Median"),
+	QT_TRANSLATE_NOOP("StatsTranslations", "Mean"),
+	QT_TRANSLATE_NOOP("StatsTranslations", "Time-weighted mean"),
+	QT_TRANSLATE_NOOP("StatsTranslations", "Sum")
+};
+
+QStringList StatsVariable::supportedOperationNames() const
+{
+	std::vector<StatsOperation> ops = supportedOperations();
+	QStringList res;
+	res.reserve(ops.size());
+	for (StatsOperation op: ops)
+		res.push_back(operationName(op));
+	return res;
+}
+
+StatsOperation StatsVariable::idxToOperation(int idx) const
+{
+	std::vector<StatsOperation> ops = supportedOperations();
+	if (ops.empty()) {
+		qWarning("Stats variable %s does not support operations", qPrintable(name()));
+		return StatsOperation::Median; // oops!
+	}
+	return idx < 0 || idx >= (int)ops.size() ? ops[0] : ops[idx];
+}
+
+QString StatsVariable::operationName(StatsOperation op)
+{
+	int idx = (int)op;
+	return idx < 0 || idx >= (int)std::size(operation_names) ? QString()
+								 : operation_names[(int)op];
+}
+
+double StatsVariable::mean(const std::vector<dive *> &dives) const
+{
+	StatsOperationResults res = applyOperations(dives);
+	return res.isValid() ? res.mean : invalid_value<double>();
+}
+
+std::vector<StatsValue> StatsVariable::values(const std::vector<dive *> &dives) const
+{
+	std::vector<StatsValue> vec;
+	vec.reserve(dives.size());
+	for (dive *d: dives) {
+		double v = toFloat(d);
+		if (is_invalid_value(v))
+			continue;
+		vec.push_back({ v, d });
```

Comment:
I know, bike shedding, but wouldn't it be much easier to just do
```
if (!is_invalid_value(v))
        vec.push_back({ v, d });
```

Reply:
done.